### ISSUE TITLE
todo rule should not flag items that contain issue tracking information

### ISSUE THREAD

**Issue by jmecosta** [SonarOpenCommunity/sonar-cxx#1038]
enforce this kind of todo
// @ todo: ABC-102 - information about todo 

perhaps define a regular expression for allowed todo messages

**Comment by guwirth** [comment]
@jmecosta why not using https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-checks/src/main/java/org/sonar/cxx/checks/CommentRegularExpressionCheck.java?

**Comment by jmecosta** [comment]
Indeed :) I will use that one instead.  Thanks

**Comment by jmecosta** [comment]
@guwirth actually, not sure if that can be used. the regular expression rule will match when todo are found, however can it be done like, if a todo is found then needs to follow this convention. i can only configure the basic regular expression, for example:

[\\][\\]\s(todo:|fixme:|@todo:|@fixme:)\s(TTT-\d{1,6}|TTSD-\d{1,6}|#\d{1,6})\s

but this will flag actually errors that are valid. so i would like to have in first instance:

[\\][\\]\s(todo|fixme|@todo|@fixme)

then apply the other regx and raise issue if not match

**Comment by guwirth** [comment]
@jmecosta I understood:

sample is:
```C++
// @ todo: ABC-102 - information about todo
```
regex: `[\][\]\s(todo:|fixme:|@todo:|@fixme:)\s(TTT-\d{1,6}|TTSD-\d{1,6}|#\d{1,6})\s`
results in false positives: ? => Can you give a sample for a false positive

Why is `[\][\]\s(todo|fixme|@todo|@fixme)` alone not a hit? Why do you need a second regex?

**Comment by jmecosta** [comment]
hum, `[\][\]\s(todo|fixme|@todo|@fixme)` will basically match all the todos, the ones with and without issue tracking information.
`[\][\]\s(todo:|fixme:|@todo:|@fixme:)\s(TTT-\d{1,6}|TTSD-\d{1,6}|#\d{1,6})\s` alone will match only the ones we want are valid. 

so we need at first instance use `[\][\]\s(todo|fixme|@todo|@fixme)` to detect all the todos, and use `[\][\]\s(todo:|fixme:|@todo:|@fixme:)\s(TTT-\d{1,6}|TTSD-\d{1,6}|#\d{1,6})\s` has a invert regx to flag lines that dont match that.

not sure if able to explain this correctly, the regx rule has a invert reg ex. But its not clear what it does, maybe its exactly what im after.

**Comment by jmecosta** [comment]
hum, https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-checks/src/main/java/org/sonar/cxx/checks/LineRegularExpressionCheck.java#L117 i think it does what i need. might be a good idea to improve the documentation of the rule?

**Comment by jmecosta** [comment]
@guwirth  hum, it does not do what i want :(. the invert property is actually weird. What are the use cases for that. even the test that we have is a bit irrealist. why would we check in real life that a line is not commented. That makes sence when for example in a specific section of code something follows  a regx but not for the all file.

or am i missing something? 

so as far as i understand we dont have in place nothing that will cover my use case?

**Comment by guwirth** [comment]
Hi, it's just:

if path is ...
if path is not ...
if comment matches regex...
if comment does not match regex ...

Regards

**Comment by jmecosta** [comment]
how?
https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-checks/src/test/java/org/sonar/cxx/checks/LineRegularExpressionCheckTest.java#L58 detects a line wihtout comment, it does not detect if comment does not match regx. 

See source code: https://github.com/SonarOpenCommunity/sonar-cxx/blob/master/cxx-checks/src/test/resources/checks/LineRegExInvert.cc#L3

**Comment by guwirth** [comment]
@jmecosta have to check the code and sample.
But do you understand / agree with how it should work? Sometimes it's easier to define a file pattern or regex how it should not be.

**Comment by jmecosta** [comment]
yes, that logic is correct. but in this case we are interested in running the regx in all source files. for example we use a rule to check that in all interface header boost must not be included and the file pattern and the regex work well there

**Comment by jmecosta** [comment]
@guwirth likely can be improved however using "negative lookahead" i am able to do what i want with a single regular expression:

```[\/][\/]( todo: | fixme: | @todo: | @fixme: )(?!TTT-\d{1,6} |TTSD-\d{1,6} |#\d{1,6} )```

pointing out just that i dont understand much about regular expressions :)
